src/main/java/org/springframework/data/couchbase/repository/query/StringBasedN1qlQueryParser.java

```diff
@@ -127,28 +129,31 @@ public StringBasedN1qlQueryParser(String statement, CouchbaseQueryMethod queryMe
 		this.queryMethod = queryMethod;
 		this.couchbaseConverter = couchbaseConverter;
 		String collection = queryMethod.getCollection();
-		this.statementContext = createN1qlSpelValues(bucketName, collection, null, null, typeField, typeValue, false, null);
-		this.countContext = createN1qlSpelValues(bucketName, collection, null, null, typeField, typeValue, true, null);
+		this.statementContext = createN1qlSpelValues(bucketName, collection, null, null, typeField, typeValue, false, null,
+				null);
+		this.countContext = createN1qlSpelValues(bucketName, collection, null, null, typeField, typeValue, true, null,
+				null);
 		this.parsedExpression = getExpression(accessor, getParameters(accessor), null, parser, evaluationContextProvider);
 		checkPlaceholders(this.parsedExpression.toString());
 	}
 
 	public StringBasedN1qlQueryParser(String bucketName, String collection, CouchbaseConverter couchbaseConverter,
-			Class domainClass, Class resultClass, String typeField, String typeValue, String[] distinctFields) {
+			Class domainClass, Class resultClass, String typeField, String typeValue, String[] distinctFields,
+			String[] fields) {
```

Comment:
what happens if both distinctFields and fields are provided? Is OK or do we need to fail on that?

Reply:
If distinctFields[] is non-null, then it will be a "distinct" query - and if distinctFields[] is non-empty, then those fields are DISTINCT().  If distinctFields is non-null, but empty, then DISTINCT() will contain all the project fields - which could be (a) all the fields in the repository entity, (b) all the fields in an alternate entity (via as(class)); or (c) all the projected fields[]. 

So having distinctFields[] non-null and empty with fields[] specifying the desired fields gives the same result as having distinctFields[] specifying the desired fields.  Maybe not allow the first?